crates/core/src/db/datastore/locking_tx_datastore/btree_index.rs

```diff
@@ -70,11 +50,17 @@ impl<'a> Iterator for BTreeIndexRangeIter<'a> {
 
     #[tracing::instrument(skip_all)]
     fn next(&mut self) -> Option<Self::Item> {
-        if let Some(key) = self.range_iter.next() {
-            self.num_keys_scanned += 1;
-            Some(&key.row_id)
-        } else {
-            None
+        loop {
+            if let Some(inner) = self.inner.as_mut() {
+                if let Some(ptr) = inner.next() {
+                    self.num_keys_scanned += 1;
+                    return Some(ptr);
+                }
+            }
+
+            self.inner = None;
+            let (_, next) = self.outer.next()?;
+            self.inner = Some(next.iter());
```

Comment:
I find this hard to follow. Could you add comments describing the flow of execution here? Or maybe a comment showing what this would look like as a chain of iterator combinators instead of in the imperative style?